LCB: Implement record type definitions

peter-b opened the pull request:
This pull request makes some progress towards implementing record types in LCB, including the following improvements:
- the "base type" mechanism for record types has been removed
- record types with the same fields are no longer considered interchangeable
- public record type info is now included in .lci files
- LCB language spec updates and compiler frontend tests

runrevmark commented:
@peter-b: This looks sound - although I suggest we hold off merging until 8.2. As it stands you can define record types but can't actually use them! We need to sort out member access syntax (which could overload [] in the first instance) and also bridging to arrays in LCS. Obviously both of these are distinct patches (from this one). If we want member access '.' operator, then we need the patch we discussed y'day with regards disallowing '.' in identifiers in LCB.

livecodeali commented on libscript/src/script-module.cpp:
```diff
@@ -1265,8 +1264,24 @@ bool MCScriptWriteInterfaceOfModule(MCScriptModuleRef self, MCStreamRef stream)
-                        // TODO - Records not yet supported
-                        break;
+                    {
+                        auto t_record_type =
```
I think changing this `auto` to `MCScriptRecordType *` is the only change needed to get this merged.

peter-b replied:
Do we have an Official Pronouncement yet on whether it's okay to use `auto` where the type is initialisation from `static_cast<>` or similar situations where without `auto` you'd need to write out the same type in a single line of code?

There doesn't seem to be any urgency to merge this so I'm happy to wait until there's a consensus.

livecodeali commented:
@livecode-vulcan review ok 0ca3f6c

livecode-vulcan commented:
💙 review by @livecodeali ok 0ca3f6c078fa1932dc927ddb015a5096b9937974

livecode-vulcan commented:
😞 test failure 0ca3f6c078fa1932dc927ddb015a5096b9937974

* try-community-linux-x86: [pending](https://vulcan.livecode.com/builders/try-community-linux-x86/builds/760)
* try-community-mac: [failure](https://vulcan.livecode.com/builders/try-community-mac/builds/765)
* try-community-linux-x86_64: [pending](https://vulcan.livecode.com/builders/try-community-linux-x86_64/builds/758)
* try-community-emscripten: [pending](https://vulcan.livecode.com/builders/try-community-emscripten/builds/759)
* try-community-ios: [pending](https://vulcan.livecode.com/builders/try-community-ios/builds/765)
* try-community-win-x86: [pending](https://vulcan.livecode.com/builders/try-community-win-x86/builds/757)
* try-community-android-armv6: [success](https://vulcan.livecode.com/builders/try-community-android-armv6/builds/757)

peter-b commented:
@livecodeali The vulcan build failed because a variable of iteration `i` was shadowing a variable of the same name in an outer loop.  I've renamed the variable in a more explanatory fashion.

livecodeali commented:
@peter-b Oops, should have caught that one!

livecodeali commented:
@livecode-vulcan review ok 4485b7b

livecode-vulcan commented:
💙 review by @livecodeali ok 4485b7b55ed058ff9595d713752aff57cd3f250e

livecode-vulcan commented:
😎 test success 4485b7b55ed058ff9595d713752aff57cd3f250e

* try-community-linux-x86: [success](https://vulcan.livecode.com/builders/try-community-linux-x86/builds/762)
* try-community-mac: [success](https://vulcan.livecode.com/builders/try-community-mac/builds/767)
* try-community-linux-x86_64: [success](https://vulcan.livecode.com/builders/try-community-linux-x86_64/builds/760)
* try-community-emscripten: [success](https://vulcan.livecode.com/builders/try-community-emscripten/builds/761)
* try-community-ios: [success](https://vulcan.livecode.com/builders/try-community-ios/builds/767)
* try-community-win-x86: [success](https://vulcan.livecode.com/builders/try-community-win-x86/builds/759)
* try-community-android-armv6: [success](https://vulcan.livecode.com/builders/try-community-android-armv6/builds/759)

peter-b commented:
Blocked on #5043.

peter-b committed:
Add compiler tests for record definitions.

peter-b committed:
Update LCB specification to include record type definitions

peter-b committed:
lc-compile: Generate/parse record type info in interface files

peter-b committed:
Simplify LCB records.

This commit simplifies record support for LiveCode builder in two
important ways:

- Removal of the concept of a "base type" for a record.  This
  simplifies the implementation a lot, and most of the benefits of the
  base type feature can still be obtained by composition.

- Record types are no longer considered equivalent if they're made up
  of the same fields.  This is important, since it ensures that two
  records that happen to just contain a pointer don't get treated as
  interchangeable by the VM, eg

      record type RenderBuffer
         mPointer as Pointer
      end type

      record type WindowHandle
         mPointer as Pointer
      end type

  These are not interchangeable!

livecodeali commented:
@livecode-vulcan review ok 239fc80

livecode-vulcan commented:
💙 review by @livecodeali ok 239fc808da7d9c8b16fdc41eb899c308c6175cd4

livecode-vulcan commented:
😎 test success 239fc808da7d9c8b16fdc41eb899c308c6175cd4

* try-community-linux-x86: [success](https://vulcan.livecode.com/builders/try-community-linux-x86/builds/853)
* try-community-mac: [success](https://vulcan.livecode.com/builders/try-community-mac/builds/859)
* try-community-linux-x86_64: [success](https://vulcan.livecode.com/builders/try-community-linux-x86_64/builds/851)
* try-community-emscripten: [success](https://vulcan.livecode.com/builders/try-community-emscripten/builds/852)
* try-community-ios: [success](https://vulcan.livecode.com/builders/try-community-ios/builds/859)
* try-community-win-x86: [success](https://vulcan.livecode.com/builders/try-community-win-x86/builds/850)
* try-community-android-armv6: [success](https://vulcan.livecode.com/builders/try-community-android-armv6/builds/850)
